lib/spack/spack/test/concretize.py

```diff
@@ -22,190 +22,184 @@
-            self.assertTrue(name in concrete.variants)
 
-        for flag in concrete.compiler_flags.valid_compiler_flags():
-            self.assertTrue(flag in concrete.compiler_flags)
 
-        if abstract.compiler and abstract.compiler.concrete:
-            self.assertEqual(abstract.compiler, concrete.compiler)
-
-        if abstract.architecture and abstract.architecture.concrete:
-            self.assertEqual(abstract.architecture, concrete.architecture)
-
-    def check_concretize(self, abstract_spec):
-        abstract = Spec(abstract_spec)
-        concrete = abstract.concretized()
-
-        self.assertFalse(abstract.concrete)
-        self.assertTrue(concrete.concrete)
-        self.check_spec(abstract, concrete)
-
-        return concrete
-
-    def test_concretize_no_deps(self):
-        self.check_concretize('libelf')
-        self.check_concretize('libelf@0.8.13')
-
-    def test_concretize_dag(self):
-        self.check_concretize('callpath')
-        self.check_concretize('mpileaks')
-        self.check_concretize('libelf')
+def check_spec(abstract, concrete):
+    if abstract.versions.concrete:
+        assert abstract.versions == concrete.versions
+
+    if abstract.variants:
+        for name in abstract.variants:
+            avariant = abstract.variants[name]
+            cvariant = concrete.variants[name]
+            assert avariant.value == cvariant.value
+
+    if abstract.compiler_flags:
+        for flag in abstract.compiler_flags:
+            aflag = abstract.compiler_flags[flag]
+            cflag = concrete.compiler_flags[flag]
+            assert set(aflag) <= set(cflag)
+
+    for name in abstract.package.variants:
+        assert name in concrete.variants
+
+    for flag in concrete.compiler_flags.valid_compiler_flags():
+        assert flag in concrete.compiler_flags
+
+    if abstract.compiler and abstract.compiler.concrete:
+        assert abstract.compiler == concrete.compiler
+
+    if abstract.architecture and abstract.architecture.concrete:
+        assert abstract.architecture == concrete.architecture
+
+
+def check_concretize(abstract_spec):
+    abstract = Spec(abstract_spec)
+    concrete = abstract.concretized()
+    assert not abstract.concrete
+    assert concrete.concrete
+    check_spec(abstract, concrete)
+    return concrete
+
+
+@pytest.fixture(
+    params=[
+        # no_deps
+        'libelf', 'libelf@0.8.13',
+        # dag
+        'callpath', 'mpileaks', 'libelf',
+        # variant
+        'mpich+debug', 'mpich~debug', 'mpich debug=2', 'mpich',
+        # compiler flags
+        'mpich cppflags="-O3"',
+        # with virtual
+        'mpileaks ^mpi', 'mpileaks ^mpi@:1.1', 'mpileaks ^mpi@2:',
+        'mpileaks ^mpi@2.1', 'mpileaks ^mpi@2.2', 'mpileaks ^mpi@2.2',
+        'mpileaks ^mpi@:1', 'mpileaks ^mpi@1.2:2'
+    ]
+)
+def spec(request):
+    """Spec to be concretized"""
+    return request.param
+
+
+@pytest.mark.usefixtures('config', 'builtin_mock')
```

Comment:
Earlier these tests were referencing "config" defined in conftest in an implicit manner. I think this was updated in your most recent commit (i.e. this line is new). Is there anything preventing a user from referencing "config" in an implicit manner elsewhere? Brief reading suggests this can be prevented although I'm having trouble seeing how that is achieved.